### PR TITLE
Java auto instr operator

### PR DESCRIPTION
**Pull request by piotr-sumo**
###### Description
This is the first version of java auto instrumentation operator. 
In order to collect traces one of the following collectors must be installed. Please see `examples` directory for collector configurations (`jaeger.yaml` and `otel-col.yaml`):
- jaeger deployed, or
- open telemetry collector
OpenTelemetry jars version v0.3.0 beta from [here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/releases/tag/v0.3.0). Currently the operator mounts jars from host directory `/home/vagrant/ot-jars` but one can edit the `getOtJarsVolume` function to use EC2 or other cloud provider volume.

Currently only deployments are supported. In order to auto instrument the application, its deployment must have the following labels (see `java-app-otlp.yaml`):
```
"should-auto-instrument": "true"
"auto-instrumentation-exporter": "otlp"
"auto-instr-service-name": "java-app-otlp-exp"
```

###### Testing performed
- [x] This PR doesn't affect current content of this repo
- [x] `cd java-auto-instrumetation && go test ./...`

### CODE REVIEW

**Comment by sumo-drosiek** [comment]
I believe this should be moved to new repository
cc: @perk-sumo @frankreno @pmaciolek

**Comment by frankreno** [comment]
I agree with @sumo-drosiek on this.  While it is possible in the future we create an operator for core k8s collection, it may be very different than what exists today so I would propose we keep this in a new GitHub repo.

File: java-auto-instrumetation/cmd/manager/main.go [hunk 1]
```diff
@@ -0,0 +1,211 @@
+package main
+
+import (
```
Comment (pmm-sumo) [inline java-auto-instrumetation/cmd/manager/main.go:3]: In principle, there should be 3 sections in the import:
- system libraries
- external packages
- internal packages

File: java-auto-instrumetation/cmd/manager/main.go [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+	}
+
+	// Get a config to talk to the apiserver
+	cfg, err := config.GetConfig()
+	if err != nil {
+		log.Error(err, "")
+		os.Exit(1)
+	}
+
+	ctx := context.TODO()
```
Comment (pmm-sumo) [inline java-auto-instrumetation/cmd/manager/main.go:87]: I believe in such case `Background()` would be [a better choice](https://golang.org/src/context/context.go?s=7426:7451#L198)

File: java-auto-instrumetation/cmd/manager/main.go [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+	// implementing the logr.Logger interface. This logger will
+	// be propagated through the whole operator, generating
+	// uniform and structured logs.
+	logf.SetLogger(zap.Logger())
+
+	printVersion()
+
+	namespace, err := k8sutil.GetWatchNamespace()
+	if err != nil {
+		log.Error(err, "Failed to get watch namespace")
```
Comment (pmm-sumo) [inline java-auto-instrumetation/cmd/manager/main.go:76]: This two function calls can be shortened to https://golang.org/pkg/log/#Fatalf

File: java-auto-instrumetation/deploy/role.yaml [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  - update
+  - watch
+- apiGroups:
+  - apps
+  resources:
+  - deployments
+  - daemonsets
+  - replicasets
+  - statefulsets
+  verbs:
```
Comment (pmm-sumo) [inline java-auto-instrumetation/deploy/role.yaml:33]: I think this is a pretty long list of actions. I am wondering if we can limit to only those that are needed

File: java-auto-instrumetation/pkg/apis/javaautoinstr/v1alpha1/javaautoinstrumentation_types.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
```
Comment (pmm-sumo) [inline java-auto-instrumetation/pkg/apis/javaautoinstr/v1alpha1/javaautoinstrumentation_types.go:7]: Should we keep this comment? :)

File: java-auto-instrumetation/pkg/controller/javaautoinstrumentation/javaautoinstrumentation_controller.go [hunk 6]
```diff
@@ -0,0 +1,390 @@
+package javaautoinstrumentation
+
+import (
+	"context"
+	"github.com/go-logr/logr"
+	appv1 "k8s.io/api/apps/v1"
+	"strings"
+	"time"
+
+	javaautoinstrv1alpha1 "github.com/SumoLogic/sumologic-kubernetes-collection/java-auto-instrumentation/pkg/apis/javaautoinstr/v1alpha1"
```
Comment (pmm-sumo) [inline java-auto-instrumetation/pkg/controller/javaautoinstrumentation/javaautoinstrumentation_controller.go:10]: This import should be at the bottom of the imports, in separate section.

The non-systems ones from the top should come here

File: java-auto-instrumetation/pkg/controller/javaautoinstrumentation/javaautoinstrumentation_controller.go [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileJavaAutoInstrumentation) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	now := time.Now().Format(time.RFC3339)
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name,
+		"Timestamp", now)
+	reqLogger.Info("Reconciling JavaAutoInstrumentation")
+
+	// Fetch the JavaAutoInstrumentation instance
```
Comment (pmm-sumo) [inline java-auto-instrumetation/pkg/controller/javaautoinstrumentation/javaautoinstrumentation_controller.go:108]: We don't need that section, do we?

**Comment by piotr-sumo** [comment]
create repo - "sumo logic auto instrumentation operator"

File: java-auto-instrumetation/cmd/manager/main.go (hunk 1 above)
Comment (piotr-sumo) [reply java-auto-instrumetation/cmd/manager/main.go:3]: 👍

File: java-auto-instrumetation/cmd/manager/main.go (hunk 2 above)
Comment (piotr-sumo) [reply java-auto-instrumetation/cmd/manager/main.go:87]: 👍

File: java-auto-instrumetation/deploy/role.yaml (hunk 4 above)
Comment (piotr-sumo) [reply java-auto-instrumetation/deploy/role.yaml:33]: I've created https://jira.kumoroku.com/jira/browse/TRACING-705 for that

File: java-auto-instrumetation/pkg/apis/javaautoinstr/v1alpha1/javaautoinstrumentation_types.go (hunk 5 above)
Comment (piotr-sumo) [reply java-auto-instrumetation/pkg/apis/javaautoinstr/v1alpha1/javaautoinstrumentation_types.go:7]: 👍

File: java-auto-instrumetation/pkg/controller/javaautoinstrumentation/javaautoinstrumentation_controller.go (hunk 6 above)
Comment (piotr-sumo) [reply java-auto-instrumetation/pkg/controller/javaautoinstrumentation/javaautoinstrumentation_controller.go:10]: 👍

File: java-auto-instrumetation/pkg/controller/javaautoinstrumentation/javaautoinstrumentation_controller.go (hunk 7 above)
Comment (piotr-sumo) [reply java-auto-instrumetation/pkg/controller/javaautoinstrumentation/javaautoinstrumentation_controller.go:108]: 👍

**Comment by mat-rumian** [comment]
It would be very nice to see the guide - How to deploy operator and what kind of changes are needed to be done in the application deployments?

**Comment by sumo-drosiek** [comment]
Closing due to https://github.com/SumoLogic/java-auto-instrumentation-operator

**Pull request closed without merging**